### ISSUE TITLE
BO: Unable to update "Price per unit" field after importing products & combinations

### ISSUE THREAD

**Issue by ruinobreferreira** [PrestaShop/PrestaShop#15785]
Unable to update the "Price per unit" in the backoffice.

**To Reproduce**

1. Go to the import Page
2. Import products with a base price & unity base price
[Downloads.zip](https://github.com/PrestaShop/PrestaShop/files/3686303/Downloads.zip)

3. Import combinations for those products
1. Go to Catalog > Products.

2. Click on the "Pricing" tab

3. Edit "Price per unit (tax excl.)"

4. Click on "Save"

5. Refresh the page and the field "Price per unit (tax excl.)" is 0.000000

**Additional information**

PrestaShop version: 1.7.6.1
PHP version: 7.2.22

**Comment by khouloudbelguith** [comment]
Hi @nobreferreira,

Unity must be not empty.
![image](https://user-images.githubusercontent.com/16067358/66029040-54f5e680-e4f6-11e9-9ab2-3cf884639ddf.png)

Thanks to check & feedback.

**Comment by ruinobreferreira** [comment]
Hi @khouloudbelguith,

in the case I reported the field "unity" was not empty.

**Comment by khouloudbelguith** [comment]
@nobreferreira, I did not manage to reproduce the issue with PS1.7.6.1.
I attached a screen record
https://drive.google.com/file/d/1F8ZekzhUJ0OSTl2h4muNcPLJaYA5JcYt/view
Thanks to check & feedback.

**Comment by ruinobreferreira** [comment]
@khouloudbelguith 

The issue is on product update. Please check screencast below.

https://drive.google.com/file/d/1kNLuxFz75PO5tyU7aRW_nvS5QAjPkvLk/view

Quick reminder, screencastify doesn't show right button click each time I ask for a page reload.

**Comment by khouloudbelguith** [comment]
@nobreferreira, it is OK on my side.
https://drive.google.com/file/d/1mGXNqOOLJP2dLgBepRAptxVLi1Pnaq9N/view
We need to retrieve the PHP error log and the debug mode report in order to find out what's wrong.
Don't you know how to get this information? Please read the following article:
http://build.prestashop.com/howtos/misc/how-to-create-bug-report/

Thanks!

**Comment by ruinobreferreira** [comment]
Hi again @khouloudbelguith

First we import the products then the combinations. (Files attached)

- Go to "Catalog > Products"
- Click on the "Pricing" tab
- Update "Price per unit (tax excl.)" field
- Click on "Save"
- Refresh and the "Price per unit (tax excl.)" shoud be 0.000000

As attach goes the dump from the profiler. The post I did is on "step 2" page 8 with unit price "0.28"

Hope it helps.

If needed we can arrange an access.

[Downloads.zip](https://github.com/PrestaShop/PrestaShop/files/3681538/Downloads.zip)
[ProductController.pdf](https://github.com/PrestaShop/PrestaShop/files/3681539/ProductController.pdf)

**Comment by khouloudbelguith** [comment]
@nobreferreira, I manage to reproduce the issue with PS1.7.6.1 & PS1.7.5.2 after importing the products & combinations. 

https://drive.google.com/file/d/1zVQzcTNbarrvX_KEpzByq6olCyv47yrs/view
I’ll add this to the debug roadmap so that it’s fixed. If you have already fixed it on your end or if you think you can do it, please do send us a pull request!
Without import => OK
Thanks!

**Comment by flambert860** [comment]
Will there soon be a fix for this or workaround?

**Comment by 3sal** [comment]
Is there any update or anyone can direct me to fix available?

**Comment by SuchyTheMaster** [comment]
I have the same problem. 

- Prestashop 1.7.7.6 with multistore.
- PHP 7.3 fpm + IonCube + nginx
- disabled cache, memcached
- forced smarty rendering (disabled smarty cache)


I've noticed those facts:

1. It happens only when products have combinations and were imported (imported simple products or products with combinations bu created directly in admin panel are ok)
2. Wrong products has set `unit_price_ratio` to 0 in table `ps_products` and `ps_products_shop`
3. When I change those field directly in database, then price for unit is correct, but till the first saving product in admin panel (even when I don't change anything). After saving base product for combinations has unit_price_ratio set to 0.
4. Both good and "broken" products have the same saving requests (except different id_product of course ;)). But when I save those "broken" product then in saving response is `unit_price_ratio: "0.000000"`, but in good product (with combination but not imported) in sponse exist `unit_price_ratio` > 0 (for example 1.000000). 

My workaround

1. Because I have constant value `unit_price_ratio` for specific `unity` (for example unity `kg` should give me `unit_price_ratio = 0,001` because i have selling price for 1 g
2. So I've made trigger in database which sets this value in `unit_price_ratio` when original value is 0 an unit equals `kg`
3. This solution have disadvantages

- block setting 0 as unit price for specified unit
- block setting different unit ratio values for the same unity
- adds load on database
 
SQL query with create trigges

```
DELIMITER $$
CREATE TRIGGER `unityfix_ps_product_shop_update` 
BEFORE UPDATE ON `ps_product_shop`
 FOR EACH ROW BEGIN
    IF NEW.unity = 'kg' THEN
    	SET NEW.unit_price_ratio = 0.001;
    END IF;
END$$

CREATE TRIGGER `unityfix_ps_product_update` 
BEFORE UPDATE ON `ps_product`
 FOR EACH ROW BEGIN
    IF NEW.unity = 'kg' THEN
    	SET NEW.unit_price_ratio = 0.001;
    END IF;
END$$
DELIMITER ;
```

If You need other price ratios for other unity value, then just extend IF - END IF block;

**This is only workaround! You should update Your Presta when Preta Team will fix this in code, and remove triggers**

**Comment by PawelRanuszkiewicz** [comment]
I've diagnosed the problem down to AdminProductWrapper.php

Commenting out lines 210-217 (starting at Product::disableCache() and ending at Product::enableCache()) fixes the issue.
I believe the problem comes from the cache mechanism somehow. Maybe products that are imported somehow aren't cached? And the cached "price per unit" overwrites the form input one?

**Comment by Triloworld** [comment]
Yes. Problem after combination only import and with 1.7.7.0 version
Link to what @PawelRanuszkiewicz (thx!) refer: https://github.com/PrestaShop/PrestaShop/blob/ab9309a5d45c0dcb56e7885173e283e4a8ad005d/src/Adapter/Product/AdminProductWrapper.php#L207-L214

Commenting only cache change won't resolve solution.
**Workaround**:
We comment it and is working now (be sure that import don't require setAvailableDate method (your import don't have any Available Date set).

**Comment by kpodemski** [comment]
@MatShir to be tested on 8

**Comment by MatShir** [comment]
Seems to be fixed on 8.1 😮

**Comment by hibatallahAouadni** [comment]
Fixed with PS810
